src/Shared/WeakStringCache.cs

```diff
@@ -0,0 +1,145 @@
+    {
+        /// <summary>
+        /// Debug stats returned by GetDebugInfo().
+        /// </summary>
+        public struct DebugInfo
+        {
+            public int LiveStringCount;
+            public int CollectedStringCount;
+        }
+
+        /// <summary>
+        /// Holds a weak GC handle to a string. Shared by all strings with the same hash code and referencing the last such string we've seen.
+        /// </summary>
+        private struct StringWeakHandle
+        {
+            /// <summary>
+            /// Weak GC handle to the last string of the given hashcode we've seen.
+            /// </summary>
+            public GCHandle WeakHandle;
+
+            /// <summary>
+            /// Returns true if the string referenced by the handle is still alive.
+            /// </summary>
+            public bool IsUsed => WeakHandle.Target != null;
+
+            /// <summary>
+            /// Returns the string referenced by this handle if it is equal to the given internable.
+            /// </summary>
+            /// <param name="internable">The internable describing the string we're looking for.</param>
+            /// <returns>The string matching the internable or null if the handle is referencing a collected string or the string is different.</returns>
+            public string GetString<T>(T internable) where T : IInternable
+            {
+                if (WeakHandle.IsAllocated && WeakHandle.Target is string str)
+                {
+                    if (internable.Length == str.Length &&
+                        internable.StartsWithStringByOrdinalComparison(str))
+                    {
+                        return str;
+                    }
+                }
+                return null;
+            }
+
+            /// <summary>
+            /// Sets the handle to the given string. If the handle is still referencing another live string, that string is effectively forgotten.
+            /// </summary>
+            /// <param name="str">The string to set.</param>
+            public void SetString(string str)
+            {
+                if (!WeakHandle.IsAllocated)
+                {
+                    // The handle is not allocated - allocate it.
+                    WeakHandle = GCHandle.Alloc(str, GCHandleType.Weak);
+                }
+                else
+                {
+                    WeakHandle.Target = str;
+                }
+            }
+
+            /// <summary>
+            /// Frees the GC handle.
+            /// </summary>
+            public void Free()
+            {
+                WeakHandle.Free();
+            }
+        }
+
+        /// <summary>
+        /// Improvised capacity for the scavenging heuristics. We use this number as the initial capacity of the underlying collection
+        /// and then as the maximum size we let the collection grow before scavenging unused entries.
+        /// </summary>
+        private int _capacity = 503;
```

Comment:
nit: is it worth renaming this to something like `_scavengeThreshold`?

Reply:
Yes, this was confusing. I have added a new const `_initialCapacity` and renamed the r/w field to `_scavengeThreshold` as suggested. Thank you!